src/device-config.ts

```diff
@@ -0,0 +1,512 @@
+import { EnvVarObject } from './lib/types';
+import { checkInt, checkTruthy } from './lib/validation';
+import { DeviceApplicationState } from './types/state';
+
+const vpnServiceName = 'openvpn-resin';
+
+interface DeviceConfigConstructOpts {
+	db: Database;
+	config: Config;
+	logger: Logger;
+}
+
+interface ConfigOption {
+	envVarName: string;
+	varType: string;
+	defaultValue: string;
+	rebootRequired?: boolean;
+}
+
+interface ConfigStep {
+	// TODO: This is a bit of a mess, the DeviceConfig class shouldn't
+	// know that the reboot action exists as it is implemented by
+	// DeviceState. Fix this weird circular dependency
+	action: keyof DeviceActionExecutors | 'reboot';
+	humanReadableTarget?: Dictionary<string>;
+	target?: string | Dictionary<string>;
+	rebootRequired?: boolean;
+}
+
+interface DeviceActionExecutorOpts {
+	initial?: boolean;
+}
+
+type DeviceActionExecutorFn = (
+	step: ConfigStep,
+	opts?: DeviceActionExecutorOpts,
+) => Promise<void>;
+
+interface DeviceActionExecutors {
+	changeConfig: DeviceActionExecutorFn;
+	setVPNEnabled: DeviceActionExecutorFn;
+	setBootConfig: DeviceActionExecutorFn;
+}
+
+export class DeviceConfig {
+	private db: Database;
+	private config: Config;
+	private logger: Logger;
+	private rebootRequired = false;
+	private actionExecutors: DeviceActionExecutors;
+	private configBackend: DeviceConfigBackend | null = null;
+
+	private static configKeys: Dictionary<ConfigOption> = {
+		appUpdatePollInterval: {
+			envVarName: 'SUPERVISOR_POLL_INTERVAL',
+			varType: 'int',
+			defaultValue: '60000',
+		},
+		localMode: {
+			envVarName: 'SUPERVISOR_LOCAL_MODE',
+			varType: 'bool',
+			defaultValue: 'false',
+		},
+		connectivityCheckEnabled: {
+			envVarName: 'SUPERVISOR_CONNECTIVITY_CHECK',
+			varType: 'bool',
+			defaultValue: 'true',
+		},
+		loggingEnabled: {
+			envVarName: 'SUPERVISOR_LOG_CONTROL',
+			varType: 'bool',
+			defaultValue: 'true',
+		},
+		delta: {
+			envVarName: 'SUPERVISOR_DELTA',
+			varType: 'bool',
+			defaultValue: 'false',
+		},
+		deltaRequestTimeout: {
+			envVarName: 'SUPERVISOR_DELTA_REQUEST_TIMEOUT',
+			varType: 'int',
+			defaultValue: '30000',
+		},
+		deltaApplyTimeout: {
+			envVarName: 'SUPERVISOR_DELTA_APPLY_TIMEOUT',
+			varType: 'int',
+			defaultValue: '',
+		},
+		deltaRetryCount: {
+			envVarName: 'SUPERVISOR_DELTA_RETRY_COUNT',
+			varType: 'int',
+			defaultValue: '30',
+		},
+		deltaRetryInterval: {
+			envVarName: 'SUPERVISOR_DELTA_RETRY_INTERVAL',
+			varType: 'int',
+			defaultValue: '10000',
+		},
+		deltaVersion: {
+			envVarName: 'SUPERVISOR_DELTA_VERSION',
+			varType: 'int',
+			defaultValue: '2',
+		},
+		lockOverride: {
+			envVarName: 'SUPERVISOR_OVERRIDE_LOCK',
+			varType: 'bool',
+			defaultValue: 'false',
+		},
+		persistentLogging: {
+			envVarName: 'SUPERVISOR_PERSISTENT_LOGGING',
+			varType: 'bool',
+			defaultValue: 'false',
+			rebootRequired: true,
+		},
+	};
+
+	static validKeys = ['SUPERVISOR_VPN_CONTROL', 'OVERRRIDE_LOCK'].concat(
```

Comment:
You could make this:
```suggestion
	static validKeys = ['SUPERVISOR_VPN_CONTROL', 'OVERRRIDE_LOCK', ..._.map(DeviceConfig.configKeys, 'envVarName')]
```